### PR TITLE
WIP: adding PV robust version of DeepMET

### PR DESCRIPTION
**Pull request by yongbinfeng**
PR to add the PV robust version of DeepMET into the main production branch. 

Added two new set of variables now:
(`DeepMETPVRobust_pt`, `DeepMETPVRobust_phi`)
(`DeepMETPVRobustNoPUPPI_pt`, `DeepMETPVRobustNoPUPPI_phi`)
Also save a new variable `PVRobustIndex` saving the PV index closest to the leading loose muon. 

Work in progress for the validation of performance.

### CODE REVIEW

File: RecoMET/METPUSubtraction/plugins/DeepMETPVRobustProducer.cc [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        px_leptons += pf.px();
+        py_leptons += pf.py();
+        continue;
+      }
+    }
+
+    // fill the tensor
+    // dz, eta, mass, pt, puppi, px, py
+    float* ptr = &input_.tensor<float, 3>()(0, i_pf, 0);
+    *ptr = pf.dz();
```
Comment (bendavid) [inline RecoMET/METPUSubtraction/plugins/DeepMETPVRobustProducer.cc]: Doesn't this just compute delta z with respect to the first primary vertex in the collection? (or is that already set to the one closest to the muon at this stage?)

Comment (yongbinfeng) [reply RecoMET/METPUSubtraction/plugins/DeepMETPVRobustProducer.cc]: working on propogating the updated information to the DeepMET producer currently (updates soon).

maybe take a look at https://github.com/WMass/cmssw/pull/33/files#diff-9eb7cb44e20caecc7eb7b328c9c2ee434995e8991a6f07d5eb231417b9ad50c5R97-R139 this part first and see if the PV determinition looks fine?

File: CommonTools/PileupAlgos/plugins/PuppiPVRobustProducer.cc [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+  //double pv_z = 0;
+  for (auto const& muon : *muonCol) {
+    if (muon.pt() < 5)
+      continue;
+    if (!muon.isLooseMuon())
+      continue;
+
+    //std::cout << "muonz " << muon.vz() << " BestTrack_z " << muon.muonBestTrack()->vz() << std::endl;
+    // is this the best way to get the z position of the muon?
+    muonz = muon.vz();
```
Comment (bendavid) [inline CommonTools/PileupAlgos/plugins/PuppiPVRobustProducer.cc]: Safer would be muon.muonBestTrack()->dz(beamPoint) + beamPoint.z()

**Comment by yongbinfeng** [comment]
Running on small set of MC samples for validating the performance, but code should basically be ready for review.

**Comment by bendavid** [comment]
Would it be possible to also store the index of the muon which has been used to select the vertex and/or create a fake one?

**Comment by yongbinfeng** [comment]
> Would it be possible to also store the index of the muon which has been used to select the vertex and/or create a fake one?

done. added a new variable `PVMuonIndex` that saves the index of the muon in the slimmedMuons collection that is used to determin the "robust" PV.

the index corresponds to the index in the slimmedMuon collection though, not always the muon index in the flat ntuples (some muon quality selections applied before making flat tables). Do you prefer to save that instead? Would need to change the input tag and modify the cms.sequence a bit if so.

**Comment by bendavid** [comment]
Yes it would be more useful to have the index corresponding to the nanoaod of course.

But if this can't be done quickly we can also forget about it.

Does the vertex index at least correspond to what is stored in the nano?

**Comment by yongbinfeng** [comment]
> Yes it would be more useful to have the index corresponding to the nanoaod of course.
> 
> But if this can't be done quickly we can also forget about it.

Lets merge this PR (if there is no further issue) and start the production for now. I can make a PR later once we have a fix on this? It is not very complicated in logic, but requires some code change and a bit validation. 

> 
> Does the vertex index at least correspond to what is stored in the nano?

Yes, i did not do the full validation, but they are all using `offlineSlimmedPreimaryVertices` without any futher quality cut: https://github.com/WMass/cmssw/pull/33/files#diff-26fcea4281fed36d2b76b5b69235cbf5b7bb69ddf96ce5e9b62e256853259b80R108 so the index should be matched.

**Pull request merged**